File: in_Python/0496 Next Greater Element I.py
class Solution(object):
    def nextGreaterElement(self, nums1, nums2):
        """
        :type nums1: List[int]
        :type nums2: List[int]
        :rtype: List[int]
        """
        stack = []
        dict = {}
        for i in nums2:
            while len(stack) != 0 and i > stack[-1]:
                dict[stack[-1]] = i
                stack.pop()
            stack.append(i)
        while len(stack) != 0:
            dict[stack[-1]] = -1
            stack.pop()            
        for i in range(len(nums1)): nums1[i] = dict[nums1[i]]
        return nums1
                
        

File: in_Java/0496 Next Greater Element I.java
class Solution {
    public int[] nextGreaterElement(int[] nums1, int[] nums2) {
        Stack<Integer> st = new Stack<>();
        int[] temp = new int[nums2.length];
        for(int i = 0; i < nums2.length; i++){
            while(!st.isEmpty() && nums2[st.peek()] < nums2[i]){
                temp[st.peek()] = nums2[i];
                st.pop(); 
            }
            st.push(i);
        }
        while(!st.isEmpty()){
            temp[st.pop()] = -1;
        }
        int[] res = new int[nums1.length];
        for(int i = 0; i < res.length; i++){
            res[i] = temp[idxof(nums2,nums1[i])];
        }
        return res;
    }
    private int idxof(int[] x, int n){
        for(int i = 0; i < x.length; i++){
            if(x[i] == n) return i;
        }
        return -1;
    }
}